In [25]:
import pandas as pd

## Reading in the Data

In [26]:
df = pd.read_csv('./datasets/training.csv')
df.head()

Engagements  Followers at Posting                  Created   Type  \
0       502093              36984682  2019-05-21 23:30:51 EDT  Video   
1       603380              36984682  2019-05-21 22:53:33 EDT  Video   
2       603380              36984682  2019-05-21 22:19:58 EDT  Video   
3       725100              36984682  2019-05-21 22:02:41 EDT  Video   
4       661446              36984682  2019-05-21 20:47:49 EDT  Video   

                                         Description  
0  The @raptors bench trio of @sergeibaka @norman...  
1  @kyle_lowry7 pulls from deep for the @raptors ...  
2     @k_mid22 with some english on the @bucks dime!  
3        Kawhi punches it home with the left on TNT!  
4  @giannis_an34 goes baseline early to rock the ...

In [27]:
df.isnull().sum()

Engagements              0
Followers at Posting     0
Created                  0
Type                     0
Description             14
dtype: int64

In [28]:
df.shape

(7766, 5)

## Dropping 14 Instagram Posts with No Description 
- For my modeling, I want to execute NLP; therefore, I'm going to drop the 14 instagrams posts that don't have descriptions. 

In [29]:
#dropping permanently 
df.dropna(subset= ['Description'], axis = 0, inplace = True)

In [30]:
df.shape

(7752, 5)

In [31]:
df.isnull().sum().sum()

0

In [32]:
df.dtypes

Engagements              int64
Followers at Posting     int64
Created                 object
Type                    object
Description             object
dtype: object

In [33]:
#want to see the frequency of different types 
df['Type'].value_counts()

Video    5451
Photo    1588
Album     713
Name: Type, dtype: int64

## Getting Dummy Columns for Type 

In [34]:
#note how i dropped one of the dummy columns to prevent multicollinearity 
dataframe = pd.get_dummies(df, columns = ['Type'], drop_first= True )

In [35]:
dataframe.head()

Engagements  Followers at Posting                  Created  \
0       502093              36984682  2019-05-21 23:30:51 EDT   
1       603380              36984682  2019-05-21 22:53:33 EDT   
2       603380              36984682  2019-05-21 22:19:58 EDT   
3       725100              36984682  2019-05-21 22:02:41 EDT   
4       661446              36984682  2019-05-21 20:47:49 EDT   

                                         Description  Type_Photo  Type_Video  
0  The @raptors bench trio of @sergeibaka @norman...           0           1  
1  @kyle_lowry7 pulls from deep for the @raptors ...           0           1  
2     @k_mid22 with some english on the @bucks dime!           0           1  
3        Kawhi punches it home with the left on TNT!           0           1  
4  @giannis_an34 goes baseline early to rock the ...           0           1

In [36]:
import seaborn as sns

In [37]:
from sklearn.model_selection import train_test_split

**The following code shows the following features I want to keep (note after the train test split, Im going to separate `Description` for both train and test in order to execute nlp to ultimately vectorize).**



**My target variable is `Engagements`**

In [38]:
features = ['Followers at Posting', 'Description', 'Type_Photo', 'Type_Video']
X = dataframe[features]
y = dataframe['Engagements']

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

**Separating `Description` from other features in order to clean text data for vectorization**

In [40]:
X_train_nlp = X_train[['Description']]
X_test_nlp = X_test[['Description']]

# Cleaning Up Train Data
- At this point, I have two text dataframes (train and test)
- The following code will demonstrate my process for cleaning each dataframe. 

In [41]:
#sanity check
X_train_nlp.head()

Description
6709  A new career-high & @pelicansnba franchise-hig...
1615    Swipe through to recap Wednesday’s seven games!
980             @dloading breaks out an array of moves!
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...
6991  @dwighthoward takes it coast to coast for the ...

In [42]:
#sanity check
X_train_nlp['Description'][6709]

'A new career-high & @pelicansnba franchise-high 25 assists for @rajonrondo tonight!'

In [43]:
import string 

## Step 1: Removing Punctuations and Numbers

In [44]:
#viewing all punctuations 
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [45]:
#creating a list of all punctuations
remove = list(string.punctuation)

In [46]:
#adding all digits to the remove list 
remove.extend(['0','1','2','3','4','5','6','7','8','9'])

In [47]:
#creating a function that will remove all punctuations and numbers 
from string import punctuation
def strip_punctuation(s):
    return ''.join(c for c in s if c not in remove)

In [48]:
#apply the function to both test and train data 
X_train_nlp['no punctuation and numbers'] = X_train_nlp['Description'].apply(strip_punctuation)
X_test_nlp['no punctuation and numbers'] = X_test_nlp['Description'].apply(strip_punctuation)

/Users/valamani/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/valamani/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [49]:
#sanity check
X_train_nlp.head()

Description  \
6709  A new career-high & @pelicansnba franchise-hig...   
1615    Swipe through to recap Wednesday’s seven games!   
980             @dloading breaks out an array of moves!   
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...   
6991  @dwighthoward takes it coast to coast for the ...   

                             no punctuation and numbers  
6709  A new careerhigh  pelicansnba franchisehigh  a...  
1615     Swipe through to recap Wednesday’s seven games  
980               dloading breaks out an array of moves  
3759     No giannisan ELECTRIFIES The Garden  BESTofNBA  
6991  dwighthoward takes it coast to coast for the N...

In [50]:
#sanity check
X_test_nlp.head()

Description  \
5087  ?? “Ain’t no time better than this” ...listen ...   
5603                    How’d @spidadmitchell do that!?   
1626  @giannis_an34 finds @malcolmbrogdon for the @b...   
6832                             @pushat305 in traffic!   
3896  10-year-old Phoenix Johnson shows off CRAZY ha...   

                             no punctuation and numbers  
5087   “Ain’t no time better than this” listen in to...  
5603                       How’d spidadmitchell do that  
1626  giannisan finds malcolmbrogdon for the bucks t...  
6832                                  pushat in traffic  
3896  yearold Phoenix Johnson shows off CRAZY handle...

## Step 2: Adding Vader Polarity Scores

In [56]:
#import and instantiate vader
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyser = SentimentIntensityAnalyzer()

In [57]:
#applying vader scores to the text without punctuations and numbers 
X_train_nlp['vader'] = [analyser.polarity_scores(text) for text in X_train_nlp['no punctuation and numbers']]
X_test_nlp['vader'] = [analyser.polarity_scores(text) for text in X_test_nlp['no punctuation and numbers']]

/Users/valamani/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/valamani/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [58]:
#sanity check
X_train_nlp.head()

Description  \
6709  A new career-high & @pelicansnba franchise-hig...   
1615    Swipe through to recap Wednesday’s seven games!   
980             @dloading breaks out an array of moves!   
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...   
6991  @dwighthoward takes it coast to coast for the ...   

                             no punctuation and numbers  \
6709  A new careerhigh  pelicansnba franchisehigh  a...   
1615     Swipe through to recap Wednesday’s seven games   
980               dloading breaks out an array of moves   
3759     No giannisan ELECTRIFIES The Garden  BESTofNBA   
6991  dwighthoward takes it coast to coast for the N...   

                                                  vader  
6709  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
1615  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
980   {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
3759  {'neg': 0.306, 'neu': 0.694, 'pos': 0.0, 'comp...  
6991  {'neg': 0.27, 'neu': 0.73, 'pos': 0.0, 'compou...

In [59]:
#sanity check
X_test_nlp.head()

Description  \
5087  ?? “Ain’t no time better than this” ...listen ...   
5603                    How’d @spidadmitchell do that!?   
1626  @giannis_an34 finds @malcolmbrogdon for the @b...   
6832                             @pushat305 in traffic!   
3896  10-year-old Phoenix Johnson shows off CRAZY ha...   

                             no punctuation and numbers  \
5087   “Ain’t no time better than this” listen in to...   
5603                       How’d spidadmitchell do that   
1626  giannisan finds malcolmbrogdon for the bucks t...   
6832                                  pushat in traffic   
3896  yearold Phoenix Johnson shows off CRAZY handle...   

                                                  vader  
5087  {'neg': 0.078, 'neu': 0.642, 'pos': 0.279, 'co...  
5603  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
1626  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
6832  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  
3896  {'neg': 0.18, 'neu': 0.688, 'pos': 0.132, 'com...

## Step 3: Pull out all Vader Scores

In [60]:
#this code creates new columns of vader scores for train
X_train_nlp['neg'] = X_train_nlp['vader'].apply(lambda x: x.get('neg'))
X_train_nlp['neu'] = X_train_nlp['vader'].apply(lambda x: x.get('neu'))
X_train_nlp['pos'] = X_train_nlp['vader'].apply(lambda x: x.get('pos'))
X_train_nlp['compound'] = X_train_nlp['vader'].apply(lambda x: x.get('compound'))

In [61]:
#this code creates new columns of vader scores for test
X_test_nlp['neg'] = X_test_nlp['vader'].apply(lambda x: x.get('neg'))
X_test_nlp['neu'] = X_test_nlp['vader'].apply(lambda x: x.get('neu'))
X_test_nlp['pos'] = X_test_nlp['vader'].apply(lambda x: x.get('pos'))
X_test_nlp['compound'] = X_test_nlp['vader'].apply(lambda x: x.get('compound'))

In [63]:
#sanity check
X_train_nlp.head()

Description  \
6709  A new career-high & @pelicansnba franchise-hig...   
1615    Swipe through to recap Wednesday’s seven games!   
980             @dloading breaks out an array of moves!   
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...   
6991  @dwighthoward takes it coast to coast for the ...   

                             no punctuation and numbers  \
6709  A new careerhigh  pelicansnba franchisehigh  a...   
1615     Swipe through to recap Wednesday’s seven games   
980               dloading breaks out an array of moves   
3759     No giannisan ELECTRIFIES The Garden  BESTofNBA   
6991  dwighthoward takes it coast to coast for the N...   

                                                  vader    neg    neu  pos  \
6709  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
1615  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
980   {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
3759  {'neg': 0.306, 'neu': 0.694, 'pos': 0.0, 'comp...  0.306  0.694  0.0   
6991  {'neg': 0.27, 'neu': 0.73, 'pos': 0.0, 'compou...  0.270  0.730  0.0   

      compound  
6709     0.000  
1615     0.000  
980      0.000  
3759    -0.296  
6991    -0.516

In [64]:
#sanity check
X_test_nlp.head()

Description  \
5087  ?? “Ain’t no time better than this” ...listen ...   
5603                    How’d @spidadmitchell do that!?   
1626  @giannis_an34 finds @malcolmbrogdon for the @b...   
6832                             @pushat305 in traffic!   
3896  10-year-old Phoenix Johnson shows off CRAZY ha...   

                             no punctuation and numbers  \
5087   “Ain’t no time better than this” listen in to...   
5603                       How’d spidadmitchell do that   
1626  giannisan finds malcolmbrogdon for the bucks t...   
6832                                  pushat in traffic   
3896  yearold Phoenix Johnson shows off CRAZY handle...   

                                                  vader    neg    neu    pos  \
5087  {'neg': 0.078, 'neu': 0.642, 'pos': 0.279, 'co...  0.078  0.642  0.279   
5603  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
1626  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
6832  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
3896  {'neg': 0.18, 'neu': 0.688, 'pos': 0.132, 'com...  0.180  0.688  0.132   

      compound  
5087    0.7672  
5603    0.0000  
1626    0.0000  
6832    0.0000  
3896   -0.2103

## Step 4: Tokenize then Lemmatize 
- I decided to proceed with lemmatizing since it spits out the base or dictionary form of the word. 

In [65]:
import nltk

In [66]:
from nltk.tokenize import RegexpTokenizer
from nltk.stem import WordNetLemmatizer

In [67]:
#instntiate tokenizer
tokenizer = RegexpTokenizer(r'\w+')
lem = WordNetLemmatizer()

In [68]:
#creating a lemmatize function that tokenizes then lemmatize the test
def lemmatize(text):
    return [lem.lemmatize(w.lower()) for w in tokenizer.tokenize(text)]  



In [69]:
#let's apply that function to both train and test 
X_train_nlp['lemmatized'] = X_train_nlp['no punctuation and numbers'].apply(lemmatize)
X_test_nlp['lemmatized'] = X_test_nlp['no punctuation and numbers'].apply(lemmatize)

In [70]:
#sanity check
X_train_nlp.head()

Description  \
6709  A new career-high & @pelicansnba franchise-hig...   
1615    Swipe through to recap Wednesday’s seven games!   
980             @dloading breaks out an array of moves!   
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...   
6991  @dwighthoward takes it coast to coast for the ...   

                             no punctuation and numbers  \
6709  A new careerhigh  pelicansnba franchisehigh  a...   
1615     Swipe through to recap Wednesday’s seven games   
980               dloading breaks out an array of moves   
3759     No giannisan ELECTRIFIES The Garden  BESTofNBA   
6991  dwighthoward takes it coast to coast for the N...   

                                                  vader    neg    neu  pos  \
6709  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
1615  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
980   {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
3759  {'neg': 0.306, 'neu': 0.694, 'pos': 0.0, 'comp...  0.306  0.694  0.0   
6991  {'neg': 0.27, 'neu': 0.73, 'pos': 0.0, 'compou...  0.270  0.730  0.0   

      compound                                         lemmatized  
6709     0.000  [a, new, careerhigh, pelicansnba, franchisehig...  
1615     0.000  [swipe, through, to, recap, wednesday, s, seve...  
980      0.000        [dloading, break, out, an, array, of, move]  
3759    -0.296  [no, giannisan, electrifies, the, garden, best...  
6991    -0.516  [dwighthoward, take, it, coast, to, coast, for...

In [71]:
#sanity check
X_test_nlp.head()

Description  \
5087  ?? “Ain’t no time better than this” ...listen ...   
5603                    How’d @spidadmitchell do that!?   
1626  @giannis_an34 finds @malcolmbrogdon for the @b...   
6832                             @pushat305 in traffic!   
3896  10-year-old Phoenix Johnson shows off CRAZY ha...   

                             no punctuation and numbers  \
5087   “Ain’t no time better than this” listen in to...   
5603                       How’d spidadmitchell do that   
1626  giannisan finds malcolmbrogdon for the bucks t...   
6832                                  pushat in traffic   
3896  yearold Phoenix Johnson shows off CRAZY handle...   

                                                  vader    neg    neu    pos  \
5087  {'neg': 0.078, 'neu': 0.642, 'pos': 0.279, 'co...  0.078  0.642  0.279   
5603  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
1626  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
6832  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
3896  {'neg': 0.18, 'neu': 0.688, 'pos': 0.132, 'com...  0.180  0.688  0.132   

      compound                                         lemmatized  
5087    0.7672  [ain, t, no, time, better, than, this, listen,...  
5603    0.0000                 [how, d, spidadmitchell, do, that]  
1626    0.0000  [giannisan, find, malcolmbrogdon, for, the, bu...  
6832    0.0000                              [pushat, in, traffic]  
3896   -0.2103  [yearold, phoenix, johnson, show, off, crazy, ...

**As you can see above, the lemmatize function returns a list of words separated by commas.**

**In order to vectorize properly, we must change the lemmatize column to a string without any brackets or commas.** 

### The following code will execute the transformation to string for both test and train. 
- The final string we will use to vectorize will be called `stringOG`

In [72]:

X_train_nlp['texty'] = [[' '.join(i)] for i in X_train_nlp['lemmatized']]
X_train_nlp['string'] = [str(i) for i in X_train_nlp['texty']]
X_train_nlp['stringOG'] = [i.replace('[', '').replace(']', '').replace("'", '') for i in X_train_nlp['string']]

X_test_nlp['texty'] = [[' '.join(i)] for i in X_test_nlp['lemmatized']]
X_test_nlp['string'] = [str(i) for i in X_test_nlp['texty']]
X_test_nlp['stringOG'] = [i.replace('[', '').replace(']', '').replace("'", '') for i in X_test_nlp['string']]


In [73]:
#sanity check
X_train_nlp.head()

Description  \
6709  A new career-high & @pelicansnba franchise-hig...   
1615    Swipe through to recap Wednesday’s seven games!   
980             @dloading breaks out an array of moves!   
3759  No. 2??...@giannis_an34 ELECTRIFIES The Garden...   
6991  @dwighthoward takes it coast to coast for the ...   

                             no punctuation and numbers  \
6709  A new careerhigh  pelicansnba franchisehigh  a...   
1615     Swipe through to recap Wednesday’s seven games   
980               dloading breaks out an array of moves   
3759     No giannisan ELECTRIFIES The Garden  BESTofNBA   
6991  dwighthoward takes it coast to coast for the N...   

                                                  vader    neg    neu  pos  \
6709  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
1615  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
980   {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.0   
3759  {'neg': 0.306, 'neu': 0.694, 'pos': 0.0, 'comp...  0.306  0.694  0.0   
6991  {'neg': 0.27, 'neu': 0.73, 'pos': 0.0, 'compou...  0.270  0.730  0.0   

      compound                                         lemmatized  \
6709     0.000  [a, new, careerhigh, pelicansnba, franchisehig...   
1615     0.000  [swipe, through, to, recap, wednesday, s, seve...   
980      0.000        [dloading, break, out, an, array, of, move]   
3759    -0.296  [no, giannisan, electrifies, the, garden, best...   
6991    -0.516  [dwighthoward, take, it, coast, to, coast, for...   

                                                  texty  \
6709  [a new careerhigh pelicansnba franchisehigh as...   
1615    [swipe through to recap wednesday s seven game]   
980               [dloading break out an array of move]   
3759    [no giannisan electrifies the garden bestofnba]   
6991  [dwighthoward take it coast to coast for the n...   

                                                 string  \
6709  ['a new careerhigh pelicansnba franchisehigh a...   
1615  ['swipe through to recap wednesday s seven game']   
980             ['dloading break out an array of move']   
3759  ['no giannisan electrifies the garden bestofnba']   
6991  ['dwighthoward take it coast to coast for the ...   

                                               stringOG  
6709  a new careerhigh pelicansnba franchisehigh ass...  
1615      swipe through to recap wednesday s seven game  
980                 dloading break out an array of move  
3759      no giannisan electrifies the garden bestofnba  
6991  dwighthoward take it coast to coast for the nb...

In [74]:
#sanity check
X_test_nlp.head()

Description  \
5087  ?? “Ain’t no time better than this” ...listen ...   
5603                    How’d @spidadmitchell do that!?   
1626  @giannis_an34 finds @malcolmbrogdon for the @b...   
6832                             @pushat305 in traffic!   
3896  10-year-old Phoenix Johnson shows off CRAZY ha...   

                             no punctuation and numbers  \
5087   “Ain’t no time better than this” listen in to...   
5603                       How’d spidadmitchell do that   
1626  giannisan finds malcolmbrogdon for the bucks t...   
6832                                  pushat in traffic   
3896  yearold Phoenix Johnson shows off CRAZY handle...   

                                                  vader    neg    neu    pos  \
5087  {'neg': 0.078, 'neu': 0.642, 'pos': 0.279, 'co...  0.078  0.642  0.279   
5603  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
1626  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
6832  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...  0.000  1.000  0.000   
3896  {'neg': 0.18, 'neu': 0.688, 'pos': 0.132, 'com...  0.180  0.688  0.132   

      compound                                         lemmatized  \
5087    0.7672  [ain, t, no, time, better, than, this, listen,...   
5603    0.0000                 [how, d, spidadmitchell, do, that]   
1626    0.0000  [giannisan, find, malcolmbrogdon, for, the, bu...   
6832    0.0000                              [pushat, in, traffic]   
3896   -0.2103  [yearold, phoenix, johnson, show, off, crazy, ...   

                                                  texty  \
5087  [ain t no time better than this listen in to t...   
5603                     [how d spidadmitchell do that]   
1626  [giannisan find malcolmbrogdon for the buck tr...   
6832                                [pushat in traffic]   
3896  [yearold phoenix johnson show off crazy handle...   

                                                 string  \
5087  ['ain t no time better than this listen in to ...   
5603                   ['how d spidadmitchell do that']   
1626  ['giannisan find malcolmbrogdon for the buck t...   
6832                              ['pushat in traffic']   
3896  ['yearold phoenix johnson show off crazy handl...   

                                               stringOG  
5087  ain t no time better than this listen in to th...  
5603                       how d spidadmitchell do that  
1626  giannisan find malcolmbrogdon for the buck tri...  
6832                                  pushat in traffic  
3896  yearold phoenix johnson show off crazy handle ...

## Step 5: Vectorizer Time 

Let's only get `stringOG` from both test and train in order to vectorize. 

**Note:** I decided to use tfidf vectorizer and remove english stopwords

In [76]:
X_train_vec = X_train_nlp['stringOG']
X_test_vec = X_test_nlp['stringOG']

In [77]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [78]:
#instantiate then fit tfidf
tfidf = TfidfVectorizer(stop_words= 'english')
tfidf.fit(X_train_vec)

TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.float64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), norm='l2', preprocessor=None, smooth_idf=True,
        stop_words='english', strip_accents=None, sublinear_tf=False,
        token_pattern='(?u)\\b\\w\\w+\\b', tokenizer=None, use_idf=True,
        vocabulary=None)

## Transformation Time!
- Note how I transformed my test data based on .fit(train) in order to prevent data leakage from occuring 

In [79]:
# transoforming 
train_transform = tfidf.transform(X_train_vec)
test_transform = tfidf.transform(X_test_vec)

In [80]:
#creating df of transformed data
train_transform_df = pd.DataFrame(train_transform.toarray(),
                   columns= tfidf.get_feature_names())

test_transform_df = pd.DataFrame(test_transform.toarray(),
                   columns= tfidf.get_feature_names())

In [81]:
#sanity check
train_transform_df.shape

(5193, 3400)

## We finally made our train and test!
- Now let's append all the important features together
- The main features for both train and test are: vectorized df, vader sentiment scores, `Followers at Posting`, `Type_Photo`, and `Type_Video`. 

In [82]:
vader = ['neg', 'neu', 'pos', 'compound']
X_train_df = pd.concat(objs = [train_transform_df,
                       X_train_nlp.reset_index()[vader],
                       X_train.reset_index()[['Followers at Posting', 'Type_Photo', 'Type_Video']]], 
                       axis = 1)

In [83]:
#sanity check
X_train_df.isnull().sum().sum()

0

In [84]:
vader = ['neg', 'neu', 'pos', 'compound']
X_test_df = pd.concat(objs = [test_transform_df,
                       X_test_nlp.reset_index()[vader],
                       X_test.reset_index()[['Followers at Posting', 'Type_Photo', 'Type_Video']]], 
                       axis = 1)

In [85]:
#sanity check
X_test_df.isnull().sum().sum()

0

# Notice!
- At this point, I duplicated my notebook (called `ApplyingTEST`) in order to transform my holdout data the exact same way as the train and test data. 
- The holdout data should have 3407 columns (just like train and test). 
- Eventually after I find the best model, I will bring in the holdout data in order to predict `Engagements`

# Modeling Time!
- We finally have our dataframes for train and test
- The following is a list of models I will perform on my train and test data

## Attempting to model using multiple regressions
- Linear Regression
- Lasso
- Ridge
- Elastic
- Random Forest
- Gradient Booster
- Bernoulli
- Gaussian 
- SVR

In [86]:
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import cross_val_score

In [87]:
lr = LinearRegression()
lasso = Lasso()
ridge = Ridge()
elastic = ElasticNet()

**I am computing the cross val scores of each model to in order to pick the best performing models to ultimately perform a GridSearch.**

In [88]:
print(cross_val_score(lr, X_train_df, y_train, cv = 5).mean()) #linear regression
print(cross_val_score(lasso, X_train_df, y_train, cv = 5).mean()) #lasso
print(cross_val_score(ridge, X_train_df, y_train, cv = 5).mean()) #ridge
print(cross_val_score(elastic, X_train_df, y_train, cv = 5).mean()) #elastic

-353488.27279168036


/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


0.8612494302840685


/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=3.09316e-18): result may not be accurate.
  overwrite_a=True).T
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=3.0621e-18): result may not be accurate.
  overwrite_a=True).T
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=3.08454e-18): result may not be accurate.
  overwrite_a=True).T
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=3.13301e-18): result may not be accurate.
  overwrite_a=True).T
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=3.01971e-18): result may not be accurate.
  overwrite_a=True).T


0.8929670880868089
0.4524421300197375


In [89]:
#sanity check
X_test_df.shape

(2559, 3407)

In [90]:
#sanity check
X_train_df.head()

aarontaosunitedmasters  abc  abcespn  ability  able  academy  acfresh  \
0                     0.0  0.0      0.0      0.0   0.0      0.0      0.0   
1                     0.0  0.0      0.0      0.0   0.0      0.0      0.0   
2                     0.0  0.0      0.0      0.0   0.0      0.0      0.0   
3                     0.0  0.0      0.0      0.0   0.0      0.0      0.0   
4                     0.0  0.0      0.0      0.0   0.0      0.0      0.0   

   achievement  acknowledges  acl  ...  zhuri  zmane   zo    neg    neu  pos  \
0          0.0           0.0  0.0  ...    0.0    0.0  0.0  0.000  1.000  0.0   
1          0.0           0.0  0.0  ...    0.0    0.0  0.0  0.000  1.000  0.0   
2          0.0           0.0  0.0  ...    0.0    0.0  0.0  0.000  1.000  0.0   
3          0.0           0.0  0.0  ...    0.0    0.0  0.0  0.306  0.694  0.0   
4          0.0           0.0  0.0  ...    0.0    0.0  0.0  0.270  0.730  0.0   

   compound  Followers at Posting  Type_Photo  Type_Video  
0     0.000              26083495           0           1  
1     0.000              34492470           0           0  
2     0.000              35454842           0           1  
3    -0.296              30442523           0           1  
4    -0.516              25826105           0           1  

[5 rows x 3407 columns]

In [91]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.naive_bayes import BernoulliNB, GaussianNB

In [92]:
rf = RandomForestRegressor()
gb = GradientBoostingRegressor()
b = BernoulliNB()
g = GaussianNB()

In [93]:
print(cross_val_score(rf, X_train_df, y_train, cv = 5).mean())  #random forest
print(cross_val_score(gb, X_train_df, y_train, cv = 5).mean())   #gradient boost
print(cross_val_score(b, X_train_df, y_train, cv = 5).mean())  #bernoulli
print(cross_val_score(g, X_train_df, y_train, cv = 5).mean())   #gaussian 

/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/Users/valamani/anaconda3/lib/py

0.9040687256164937
0.883443259116367


/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


0.03139350042523398


/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


0.6240993283219586


In [94]:
from sklearn.svm import SVR

In [95]:
svr = SVR()

In [96]:
cross_val_score(svr, X_train_df, y_train, cv = 5).mean() #support vector regression 

/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/valamani/anaconda3/lib/python3.7/site-p

-0.0738795111886684

## The following models that I picked based on the cross val score are: 
|Model|Cross Val Score|
|--|--|
|Lasso|.8612|
|Ridge|.8929|
|Random Forest|.9028|
|Gradient Boosting|.8833|

In [97]:
l = Lasso()
r = Ridge()
rf = RandomForestRegressor()
gb = GradientBoostingRegressor()

## Since this competition is based on the MAPE metric, I will create a function that computes MAPE

In [98]:
#creating a mape function 
import numpy as np

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

## The following tests are modeled based on the default parameters

### Lasso

In [99]:
l.fit(X_train_df, y_train)
print(l.score(X_train_df, y_train))
print(l.score(X_test_df, y_test))

0.9638518751396787
0.8736697336135539


In [100]:
l_pred = l.predict(X_test_df)
#.predict spits out pandas series 

In [101]:
mean_absolute_percentage_error(y_test, l_pred)

8.748390118207164

### Ridge

In [102]:
r.fit(X_train_df, y_train)
print(r.score(X_train_df, y_train))
print(r.score(X_test_df, y_test))

/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: Ill-conditioned matrix (rcond=2.5934e-18): result may not be accurate.
  overwrite_a=True).T


0.9384067055012029
0.8949212583548198


In [103]:
r_pred = r.predict(X_test_df)

In [104]:
mean_absolute_percentage_error(y_test, r_pred)

8.079756724145264

### Random Forest 

In [105]:
rf.fit(X_train_df, y_train)
print(rf.score(X_train_df, y_train))
print(rf.score(X_test_df, y_test))

/Users/valamani/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9840765384886367
0.9060100190483985


In [106]:
rf_pred = rf.predict(X_test_df)

In [107]:
mean_absolute_percentage_error(y_test, rf_pred)

7.334314874977286

### Gradient Boosting

In [108]:
gb.fit(X_train_df, y_train)
print(gb.score(X_train_df, y_train))
print(gb.score(X_test_df, y_test))

0.8982712487876601
0.8829435261664901


In [109]:
gb_pred = gb.predict(X_test_df)

In [110]:
mean_absolute_percentage_error(y_test, gb_pred)

8.725065765957362

**The following table shows each models train, test, and MAPE score (based on default parameters)**

|Model|Train Score|Test Score|Mape Score|
|--|--|--|--|
|Lasso|.9638|.8736|8.749|
|Ridge|.9384|.8949|8.08|
|Random Forest|.983|.907|**7.292**|
|Gradient Boosting|.898|.882|8.719|

**Note:** My best MAPE score is 7.292

## The following code will perform GridSearch Pipeline to determine if we can get a better MAPE score. 
- Note: I didn't GridSearch Gradient Boosting since it took to long to execute. Therefore, I added ElasticNet for GridSearching. 

**The following models I performed on GridSearch are:**
- Random Forest
- Lasso
- Ridge 
- ELastic Net 

## GridSearch Pipeline Random Forest

In [111]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

In [116]:
pipe = Pipeline([
    ('rf', RandomForestRegressor())
])
pipe_params = {
    'rf__n_estimators': [500],
    'rf__max_depth': [None],
    'rf__max_features': ['auto','sqrt']
    #can add C parameter for logisitic regression
}


gs = GridSearchCV(pipe, param_grid=pipe_params, cv=3,verbose=1,n_jobs=-1)


gs.fit(X_train_df, y_train)


print(gs.best_score_)
print(f'Best Params for the Random forest are :{gs.best_params_}')

Fitting 3 folds for each of 3 candidates, totalling 9 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   4 out of   9 | elapsed:    4.4s remaining:    5.4s
[Parallel(n_jobs=-1)]: Done   9 out of   9 | elapsed:   47.9s finished


0.8696463770492735


{'rf__max_depth': None, 'rf__max_features': 'sqrt', 'rf__n_estimators': 500}

In [118]:
print(f'The train score for the gradient booster is: {gs.score(X_train_df, y_train)}')
print(f'The test score for the gradient booster is: {gs.score(X_test_df, y_test)}')
gbc_pred =gs.predict(X_test_df)
print(f'The MAPE score for the gradient booster is: {mean_absolute_percentage_error(y_test, rf_pred)}')

The train score for the gradient booster is: 0.9837459879517173
The test score for the gradient booster is: 0.878005583524015
The MAPE score for the gradient booster is: 7.334314874977286


In [121]:
last_score = 'rf__max_depth: None, rf__max_features: sqrt, rf__n_estimators: 500 with MAPE 7.33'

In [ ]:
rf_pred = gs.predict(X_test_df)

In [ ]:
mean_absolute_percentage_error(y_test, rf_pred)

## GridSearch Pipeline LASSO

In [ ]:
pipe = Pipeline([
    ('l', Lasso())
])
pipe_params = {
    'l__alpha': [0.2, 0.5, 1],
    'l__max_iter': [200, 500, 1200],
    #'rf__max_features': ['auto', None]
    #can add C parameter for logisitic regression
}


gs2 = GridSearchCV(pipe, param_grid=pipe_params, cv=3)


gs2.fit(X_train_df, y_train)


print(gs2.best_score_)
gs2.best_params_

In [ ]:
gs2.score(X_train_df, y_train)

In [ ]:
gs2.score(X_test_df, y_test)

In [ ]:
lasso_pred = gs2.predict(X_test_df)

In [ ]:

mean_absolute_percentage_error(y_test, lasso_pred)

## GridSearch Pipeline Ridge 

In [ ]:
pipe = Pipeline([
    ('r', Ridge())
])
pipe_params = {
    'r__alpha': [0.2, 0.5, 1],
    'r__max_iter': [200, 500, 1200],
    #'rf__max_features': ['auto', None]
    #can add C parameter for logisitic regression
}


gs3 = GridSearchCV(pipe, param_grid=pipe_params, cv=3)


gs3.fit(X_train_df, y_train)


print(gs3.best_score_)
gs3.best_params_

In [ ]:
gs3.score(X_train_df, y_train)

In [ ]:
gs3.score(X_test_df, y_test)

In [ ]:
ridge_pred = gs3.predict(X_test_df)
mean_absolute_percentage_error(y_test, ridge_pred)

## GridSearch ElasticNet


In [ ]:
pipe = Pipeline([
    ('e', ElasticNet())
])
pipe_params = {
    #'gb__learning_rate': [0.1],
    'e__l1_ratio': [0.9, 0.95, 1],
    #'gb__n_estimators': [100, 200, 500]
    #can add C parameter for logisitic regression
}


gs4 = GridSearchCV(pipe, param_grid=pipe_params, cv=3, n_jobs = 3)


gs4.fit(X_train_df, y_train)


print(gs4.best_score_)
gs4.best_params_

In [ ]:
print(f'The train score for the ElasticNet is: {gs4.score(X_train_df, y_train)}')
print(f'The test score for the ElasticNet is: {gs4.score(X_test_df, y_test)}')
gbc_pred =gs4.predict(X_test_df)
print(f'The MAPE score for the ElasticNet is: {mean_absolute_percentage_error(y_test, e_pred)}')

## GridSearch Gradient Booster

In [136]:
pipe = Pipeline([
    ('gbc',GradientBoostingRegressor())
])
pipe_params = {
    'gbc__n_estimators': [120,140,200],
    'gbc__max_depth': [50,80,100,200],
    'gbc__max_features': ['sqrt','log2',None]
    
}

gs5 = GridSearchCV(pipe, param_grid = pipe_params, cv=3, n_jobs=-1, verbose = 1)
gs5.fit(X_train_df,y_train)
print(f'The best score for the gradient booster is {gs5.best_score_}')

Fitting 3 folds for each of 36 candidates, totalling 108 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  34 tasks      | elapsed: 16.0min


KeyboardInterrupt: 

In [135]:
print(f'The best params for the gradient booster: {gs5.best_params_}')
print(f'The train score for the gradient booster is: {gs5.score(X_train_df, y_train)}')
print(f'The test score for the gradient booster is: {gs5.score(X_test_df, y_test)}')
gbc_pred =gs5.predict(X_test_df)
print(f'The MAPE score for the gradient booster is: {mean_absolute_percentage_error(y_test, gbc_pred)}')

The train score for the gradient booster is: 0.9981267666325299
The test score for the gradient booster is: 0.8790808456570381
The MAPE score for the gradient booster is: 8.906126800617233


In [132]:
lastbestgbs = ['MAPE 9.55, best score .844, max_features,sqrt,n_estimators 60,maxdepth 20',
'MAPE 8.9, best score ., max_features: sqrt,n_estimators:120,maxdepth:50']

## Conclusion
- Random Forest had the best overall train, test, and MAPE score when GridSearching for optimal paramters. 

Here is a table of our results when GridSearching for optimal parameters:

|Model|Train Score|Test Score|Mape Score|GridSearch Name|
|--|--|--|--|--|
|Lasso|.9638|.8736|8.749|gs2|
|Ridge|.9384|.8949|8.08|gs3|
|Random Forest|.988|.9127|**7.0428**|gs|
|Elastic Net|.963|.873|8.749|gs4|
|Gradient Booster |TBD |TBD |TBD|TBD|

**Let's bring in our test data to predict the values engagement using gs.predict. Note how the random forest is the first gridsearch; therefore, we will use gs.**

In [ ]:
#bringing in our holdout data
holdout = pd.read_csv('./datasets/X_holdout_df.csv').drop(columns = 'Unnamed: 0')
holdout.head()

In [ ]:
holdout.shape

In [ ]:
engagement_pred = gs.predict(holdout)

In [ ]:
answer = pd.DataFrame({
    'engagements': engagement_pred
})

In [ ]:
answer.head()

In [ ]:
answer.shape

In [ ]:
answer.join(holdout).head()

In [ ]:
holdout_set_erinhwang = answer.join(holdout)
holdout_set_erinhwang.head()

In [ ]:
holdout_set_erinhwang.shape

In [ ]:
holdout_set_erinhwang.isnull().sum().sum()

In [ ]:
holdout_set_erinhwang.to_csv('./datasets/holdout_set_erinhwang.csv')